Dockerfile

```diff
@@ -34,6 +36,12 @@ RUN go build -gcflags=all="-N -l" -o /usr/local/concourse/bin/concourse \
       ./cmd/concourse
 VOLUME /src
 
+
+# build the init executable for containerd
+RUN  set -x && \
+	gcc -O2 -static -o /usr/local/concourse/bin/init ./cmd/init/init.c
```

Comment:
I vaguely remember during my spike that I didn't actually need an `init` process in order to run things like `fly intercept` because containerd keeps the namespace alive and allows you to run multiple...tasks?... in it.

It looks like I did this by creating an initial task but just never running anything:

https://github.com/concourse/concourse/blob/3321b920f14acdfe12fecb87d0daec8a931233df/gcontainerd/backend.go#L190-L200

I guess this is a little hacky, but I'm just curious if that didn't work or had some weird behavior as a result of the hackiness.

Reply:
> I guess this is a little hacky, but I'm just curious if that didn't work or
> had some weird behavior as a result of the hackiness.

not that I remember :thinking:

going 1:1  with guardian, I tried following closely what they had, so it seemed
natural to go with our own init process.

howeeever, it seems like it's not really hacky to not bring our own init, even
according to the spec:


> status (string, REQUIRED) is the runtime state of the container.
> The value MAY be one of:
>
> **creating**: the container is being created (step 2 in the lifecycle) 
>
> **created**: the runtime has finished the create operation (after step 2 in the
> lifecycle), and the container process has neither exited nor executed the
> user-specified program 
>
> **running**: the container process has executed the user-specified program but
> has not exited (after step 5 in the lifecycle)
>
> see [oci/runtime.md](https://github.com/opencontainers/runtime-spec/blob/1ca19ac3226fbbba0ca30fe2d082bc0d08144152/runtime.md#state)

i.e., by not running `task.Start`, where's keeping oursselves there in the
`created` state, with `runc init` being PID 1.

which, is indeed what happens:


- in a container that has `task.Start`:

	cat /ns-proc/1/cmdline
	/tmp/gdn-init

	runc list (...)

	ID	       STATE     	BUNDLE
	b8240aff(..)   running   	/run/cont...d.runtime.v2.task/test15809...

- in a container that DOES NOT have `task.Start` issued


	cat /ns2-proc/1/cmdline | tr '\000' ' '
	runc init

	ID		STATE		BUNDLE
	a50ce69c(..)    created   	/run/cont...5f-daaf80c7ac5e...

Something that might be interesting to do then is even remove that `args`
section in the container config altogethr - from the spec's perspective, it
seems like a reasonable thing to do:

> process (object, OPTIONAL) specifies the container process. This property is
> REQUIRED **when start is called**.
>
> see [oci/config.md](https://github.com/opencontainers/runtime-spec/blob/1ca19ac3226fbbba0ca30fe2d082bc0d08144152/config.md#process)


(tl;dr: yeah, I think we can get rid of `task.Start` there and not have to bring
our own `init`, which is great!)

Reply:
@vito I was trying to see why you could get away w/out specifying `args` - as you were making the backend image aware already, it was fetching that from the image config (I think):

https://github.com/concourse/concourse/blob/3321b920f14acdfe12fecb87d0daec8a931233df/gcontainerd/backend.go#L160-L161

thus, not being empty / pointing to something nonexistent :thinking:

Reply:
(the surprising behavior of `runc` trying to `start` during a `create`):

```
START?  github.com/opencontainers/runc/libcontainer.(*linuxContainer).start(0xc0000940f0, 0xc0000b4dc0, 0x0, 0x1)
                /go/src/github.com/opencontainers/runc/libcontainer/container_linux.go:348 +0x5c7
        github.com/opencontainers/runc/libcontainer.(*linuxContainer).Start(0xc0000940f0, 0xc0000b4dc0, 0x0, 0x0)
                /go/src/github.com/opencontainers/runc/libcontainer/container_linux.go:242 +0xaf
        main.(*runner).run(0xc0001215c0, 0xc0000ee2a0, 0x0, 0x0, 0x0)
                /go/src/github.com/opencontainers/runc/utils_linux.go:320 +0xcf7
        main.startContainer(0xc0000cc9a0, 0xc0000de080, 0x1, 0x0, 0x0, 0xc0000b16b0, 0x6f4185)
                /go/src/github.com/opencontainers/runc/utils_linux.go:452 +0x43a
        main.glob..func2(0xc0000cc9a0, 0x10100, 0xc0000cc9a0)
CREATE          /go/src/github.com/opencontainers/runc/create.go:65 +0xae
        github.com/opencontainers/runc/vendor/github.com/urfave/cli.HandleAction(0x83f500, 0x8db5c0, 0xc0000cc9a0, 0xc000088700, 0x0)
                /go/src/github.com/opencontainers/runc/vendor/github.com/urfave/cli/app.go:490 +0xc8
        github.com/opencontainers/runc/vendor/github.com/urfave/cli.Command.Run(0x8c336c, 0x6, 0x0, 0x0, 0x0, 0x0, 0x0, 0x8c7a88, 0x12, 0x0, ...)
                /go/src/github.com/opencontainers/runc/vendor/github.com/urfave/cli/command.go:210 +0x991
        github.com/opencontainers/runc/vendor/github.com/urfave/cli.(*App).Run(0xc00007d1e0, 0xc000094000, 0xf, 0xf, 0x0, 0x0)
                /go/src/github.com/opencontainers/runc/vendor/github.com/urfave/cli/app.go:255 +0x6ab
        main.main()
                /go/src/github.com/opencontainers/runc/main.go:145 +0xab6
```

Reply:
https://github.com/concourse/concourse/issues/5135